conversion/conversion_test.go

```diff
@@ -0,0 +1,202 @@
+		Pks:      []ddl.IndexKey{ddl.IndexKey{Col: "col1"}},
+	}
+	return conv
+}
+
+func addForeignKeysToConv(t *testing.T, conv *internal.Conv, numFks int) {
+	var foreignKeys []ddl.Foreignkey
+	for i := 1; i <= numFks; i++ {
+		foreignKey := ddl.Foreignkey{
+			Name:         fmt.Sprintf("fk_%d", i),
+			Columns:      []string{fmt.Sprintf("col%d", i)},
+			ReferTable:   "table_b",
+			ReferColumns: []string{fmt.Sprintf("col%d", i)}}
+		foreignKeys = append(foreignKeys, foreignKey)
+	}
+	spTable := conv.SpSchema["table_a"]
+	spTable.Fks = foreignKeys
+	conv.SpSchema["table_a"] = spTable
+}
+
+func checkResults(t *testing.T, dbpath string, numFks int) {
+	ctx := context.Background()
+	resp, err := databaseAdmin.GetDatabaseDdl(ctx, &databasepb.GetDatabaseDdlRequest{Database: dbpath})
+	if err != nil {
+		t.Fatalf("Could not read DDL from database %s: %v", dbpath, err)
+	}
+
+	// Each statement in the response is the DDL for a whole table including column names, foreign key statements and primary keys.
+	// The data type is bytes. Sample resp.Statements:
+	// ["2021/07/16 13:45:14 CREATE TABLE table_a (\n  col1 STRING(10),\n  col2 STRING(10),\n) PRIMARY KEY(col1);"]
+	var stmta, stmtb string
+	if strings.Contains(resp.Statements[0], "CREATE TABLE table_a") {
+		stmta, stmtb = resp.Statements[0], resp.Statements[1]
+	} else {
+		stmta, stmtb = resp.Statements[1], resp.Statements[0]
+	}
+	assert.False(t, strings.Contains(stmtb, "FOREIGN KEY"))
+
+	wantFkStmts := []string{}
+	for i := 1; i <= numFks; i++ {
+		fkStmt := fmt.Sprintf("CONSTRAINT fk_%d FOREIGN KEY(col%d) REFERENCES table_b(col%d),", i, i, i)
+		wantFkStmts = append(wantFkStmts, fkStmt)
+	}
+	var gotFkStmts []string
+	// Filter out just the foreign key statements.
+	for _, Stmt := range strings.Split(stmta, "\n") {
+		if strings.Contains(Stmt, "FOREIGN KEY") {
+			gotFkStmts = append(gotFkStmts, strings.TrimSpace(Stmt))
+		}
+	}
+
+	sort.Strings(gotFkStmts)
+	sort.Strings(wantFkStmts)
+
+	assert.Equal(t, wantFkStmts, gotFkStmts)
+}
+
+func UpdateDDLForeignKeysUtil(t *testing.T, dbName string, numCols, numWorkers, numFks int) {
+	// Build a conv without foreign key statements to create just the tables during CreateDatabase.
+	conv := BuildConv(t, numCols)
+
+	dbpath, err := conversion.CreateDatabase(projectID, instanceID, dbName, conv, os.Stdout)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	addForeignKeysToConv(t, conv, numFks)
+	conversion.MaxWorkers = numWorkers
+	if err = conversion.UpdateDDLForeignKeys(projectID, instanceID, dbName, conv, os.Stdout); err != nil {
+		t.Fatalf("\nCan't perform update operation on db %s with foreign keys: %v\n", dbName, err)
+	}
+
+	checkResults(t, dbpath, numFks)
+	// Drop the database later.
+	defer dropDatabase(t, dbpath)
+}
+
+func TestUpdateDDLForeignKeys(t *testing.T) {
+	onlyRunForEmulatorTest(t)
+	t.Parallel()
+	UpdateDDLForeignKeysUtil(t, "test-workers-five-fks", 10, 1, 5)
+	UpdateDDLForeignKeysUtil(t, "test-workers-ten-fks", 10, 1, 10)
```

Comment:
It's more typical to write these as table driven tests. See https://github.com/cloudspannerecosystem/harbourbridge/blob/2a9101fbc6398da156ea6da2385d87e35cb0cf1a/postgres/data_test.go#L63 as an example of the same. Also https://dave.cheney.net/2019/05/07/prefer-table-driven-tests.